File: tasks/17_legacy_java_sql_repository_refactor/expected_solution/src/userRepository.ts
export interface UserSearch {
  tenantId: string;
  email?: string;
  statuses?: string[];
  includeDeleted?: boolean;
  limit: number;
}

export interface QueryPlan {
  sql: string;
  params: unknown[];
}

export function buildUserSearchQuery(search: UserSearch): QueryPlan {
  let sql = "select id,email,status from users where tenant_id = ?";
  const params: unknown[] = [search.tenantId];

  const email = search.email?.trim();
  if (email) {
    sql += " and lower(email) like ?";
    params.push(`%${email.toLowerCase()}%`);
  }
  if (search.statuses?.length) {
    sql += ` and status in (${search.statuses.map(() => "?").join(",")})`;
    params.push(...search.statuses);
  }
  if (!search.includeDeleted) {
    sql += " and deleted_at is null";
  }

  sql += " order by created_at desc limit ?";
  params.push(search.limit);
  return { sql, params };
}

File: tasks/17_legacy_java_sql_repository_refactor/expected_solution/legacy/UserRepository.java
class UserRepository {
  String search(Search s) {
    String sql = "select id,email,status from users where tenant_id = ?";
    if (s.email != null && !s.email.isBlank()) sql += " and lower(email) like ?";
    if (s.statuses != null && !s.statuses.isEmpty()) sql += " and status in (?)";
    if (!s.includeDeleted) sql += " and deleted_at is null";
    sql += " order by created_at desc limit ?";
    return sql;
  }
}
